### PR TITLE
Added Python3.9 Support

### LINKED ISSUE #888: 3.9 Support
3.9 has been out for an extended period of time, and we should see where we stand. 

Let's add 3.9 to the build matrix and get things working on 3.9.

### PR DESCRIPTION
**Pull request by daxpryce**
Python 3.9 should mostly work, but changes in the type system will necessitate some... finagling. 

In this PR we'll test this out via the build agents, which are doubtless going to fail. 

Fixes #888

### CODE REVIEW

**Comment by daxpryce** [comment]
So I think the beartype/type hinting stuff is sorted out with the latest commit - we still have broken unit tests that I will fix next, but I think the types stuff is sorted for now:

```bash
(venv) dax@cloudship graspologic: $ make fast-test | grep "PEP 484"
  /Users/dax/source/graspologic/venv/lib/python3.9/site-packages/beartype/_util/hint/pep/utilpeptest.py:373: BeartypeDecorHintPep585DeprecationWarning: PEP 484 type hint typing.List[numpy.ndarray] deprecated by PEP 585 scheduled for removal in the first Python version released after October 5th, 2025. To resolve this, either drop Python < 3.9 support and globally replace this hint by the equivalent PEP 585 type hint (e.g., "typing.List[int]" by "list[int]") or see this discussion topic for saner and more portable solutions:
  /Users/dax/source/graspologic/venv/lib/python3.9/site-packages/beartype/_util/hint/pep/utilpeptest.py:373: BeartypeDecorHintPep585DeprecationWarning: PEP 484 type hint typing.List[scipy.sparse.csr.csr_matrix] deprecated by PEP 585 scheduled for removal in the first Python version released after October 5th, 2025. To resolve this, either drop Python < 3.9 support and globally replace this hint by the equivalent PEP 585 type hint (e.g., "typing.List[int]" by "list[int]") or see this discussion topic for saner and more portable solutions:
  /Users/dax/source/graspologic/venv/lib/python3.9/site-packages/beartype/_util/hint/pep/utilpeptest.py:373: BeartypeDecorHintPep585DeprecationWarning: PEP 484 type hint typing.List[typing.Union[networkx.classes.graph.Graph, networkx.classes.ordered.OrderedGraph, networkx.classes.digraph.DiGraph, networkx.classes.ordered.OrderedDiGraph]] deprecated by PEP 585 scheduled for removal in the first Python version released after October 5th, 2025. To resolve this, either drop Python < 3.9 support and globally replace this hint by the equivalent PEP 585 type hint (e.g., "typing.List[int]" by "list[int]") or see this discussion topic for saner and more portable solutions:
  /Users/dax/source/graspologic/venv/lib/python3.9/site-packages/beartype/_util/hint/pep/utilpeptest.py:373: BeartypeDecorHintPep585DeprecationWarning: PEP 484 type hint typing.List[typing.Tuple[graspologic.pipeline.embed.embeddings.Embeddings, graspologic.pipeline.embed.embeddings.Embeddings]] deprecated by PEP 585 scheduled for removal in the first Python version released after October 5th, 2025. To resolve this, either drop Python < 3.9 support and globally replace this hint by the equivalent PEP 585 type hint (e.g., "typing.List[int]" by "list[int]") or see this discussion topic for saner and more portable solutions:
  /Users/dax/source/graspologic/venv/lib/python3.9/site-packages/beartype/_util/hint/pep/utilpeptest.py:373: BeartypeDecorHintPep585DeprecationWarning: PEP 484 type hint typing.Tuple[graspologic.pipeline.embed.embeddings.Embeddings, graspologic.pipeline.embed.embeddings.Embeddings] deprecated by PEP 585 scheduled for removal in the first Python version released after October 5th, 2025. To resolve this, either drop Python < 3.9 support and globally replace this hint by the equivalent PEP 585 type hint (e.g., "typing.List[int]" by "list[int]") or see this discussion topic for saner and more portable solutions:
  /Users/dax/source/graspologic/venv/lib/python3.9/site-packages/beartype/_util/hint/pep/utilpeptest.py:373: BeartypeDecorHintPep585DeprecationWarning: PEP 484 type hint typing.Tuple[int, int] deprecated by PEP 585 scheduled for removal in the first Python version released after October 5th, 2025. To resolve this, either drop Python < 3.9 support and globally replace this hint by the equivalent PEP 585 type hint (e.g., "typing.List[int]" by "list[int]") or see this discussion topic for saner and more portable solutions:
make: *** [fast-test] Error 1
(venv) dax@cloudship graspologic: $ git checkout 888-Python3.9
Switched to branch '888-Python3.9'
Your branch is up to date with 'origin/888-Python3.9'.
(venv) dax@cloudship graspologic: $ make fast-test | grep "PEP 484"
make: *** [fast-test] Error 1
(venv) dax@cloudship graspologic: $
```

File: graspologic/types.py [hunk 1]
```diff
@@ -5,14 +5,23 @@
...
 """
 
+import sys
 from typing import Union
 
 import networkx as nx
 import numpy as np
 import scipy.sparse as sp
 
+if sys.version_info >= (3, 9):
```
Comment (pbourke) [inline graspologic/types.py:15]: the mypy/sys.version_info insanity deserves a comment

**Review by pbourke — APPROVED** [review]
lgtm - just one comment

File: graspologic/types.py (hunk 1 above)
Comment (daxpryce) [reply graspologic/types.py:15]: something like:

```python
# mic ... not dropped, so much as thrown
```

**Comment by daxpryce** [comment]
I am not sure how to tease the bug out of this.

Our tests fail on `test_io.py` for this test:

```python
    def test_bad_delimiter(self):
        delimiter = ","
        with self.assertWarns(UserWarning):
            graphs = gs.utils.import_edgelist(self.root, delimiter=delimiter)
```

The error doesn't occur on the call to `import_edgelist`, though.  It occurs at:
```
>       with self.assertWarns(UserWarning):
E       RuntimeError: dictionary changed size during iteration
```

I was suspicious so I made the following

```python
class TestWackassWarn(unittest.TestCase):
    def test_warnings(self):
        with self.assertWarns(UserWarning):
            # unittest, I have given you the world's finest layup. Don't screw it up.
            import warnings
            warnings.warn("I am a warning")
```

It still fails:
```python
self = <tests.test_io.TestWackassWarn testMethod=test_warnings>

    def test_warnings(self):
>       with self.assertWarns(UserWarning):
E       RuntimeError: dictionary changed size during iteration

tests/test_io.py:130: RuntimeError
```

... why are you like this

**Comment by pbourke** [comment]
As discovered by @daxpryce , this was fixed for 3.6, 3.7, 3.8 and 3.10 and it looks like it didn't make it into 3.9: https://bugs.python.org/issue29620

I opened a new issue for backporting to 3.9:

https://bugs.python.org/issue45836

**Comment by daxpryce** [comment]
Re: it failing, if it's in a different module file it's fine, but something in our specific test_io module is enough to trigger the sys module path change that then triggers that error.

This is fun!

**Comment by pbourke** [comment]
yeah graspologic includes at least one module that triggers the error (IPython.utils.io). Here's a minimal case to reproduce:

```python
import unittest
import graspologic


class MyTestCase(unittest.TestCase):
    def test_something(self):
        with self.assertWarns(UserWarning):
            pass


if __name__ == '__main__':
    unittest.main()
```

**Comment by daxpryce** [comment]
`pytest.warn` doesn't have the same problem.  I spent like 3 days in an ill-conceived frenzy swapping all of our tests over to `unittest` from `pytest` because of, what I can best describe as **actual damn ghosts** in IntelliJ wasn't actually running pytests sometimes.

Turns out just restarting intellij fixed it, but I just *undid* all of the pytest testing in favor of unittest, and now I'm going back to it.  

I... I don't really want to eat crow, but I feel like there's an exceptionally compelling moral argument that I have to now

**Review by pbourke — APPROVED** [review]

**Comment by leycec** [comment]
So much thanks to @daxpryce. `unittest` fumbled the layup, but you landed seven three-pointers in a row. Larry Bird, you're on notice. :bouncing_ball_person:

**Pull request merged**